cpp/src/arrow/compute/kernels/vector_selection.cc

```diff
@@ -2355,6 +2358,76 @@ const FunctionDoc array_take_doc(
      "given by `indices`.  Nulls in `indices` emit null in the output."),
     {"array", "indices"}, "TakeOptions");
 
+const FunctionDoc indices_nonzero_doc(
+    "Return indices of the array containing non zero or false values",
+    ("For each input value, check if it's zero, false or null. Emit the index\n"
+     "of the value in the array if it's none of the those."),
+    {"values"});
+
+struct NonZeroVisitor {
+  UInt64Builder* builder;
+  const ArrayData& array;
+
+  NonZeroVisitor(UInt64Builder* builder, const ArrayData& array)
+      : builder(builder), array(array) {}
+
+  Status Visit(const DataType& type) { return Status::NotImplemented(type.ToString()); }
+
+  template <typename Type>
+  enable_if_t<is_primitive_ctype<Type>::value, Status> Visit(const Type&) {
```

Comment:
NOTE:Should we open a follow-up JIRA to support DecimalTypes?

Reply:
Please do, my main original goal was mostly to allow combining this with masks, so the Boolean type support was the most important aspect, but it definitely make sense to extend the function to support decimals.

Reply:
Create ARROW-15218 as follow-up to add Decimal support. We could also argue that any of the temporal types also fit the non-zero property, but will discuss this in JIRA.